# 🕵️‍♀️ 01. Análisis Exploratorio Preliminar (EDA)

## 📋 ¿Por qué hago esto primero?
Antes de tomar decisiones sobre cómo limpiar o transformar los datos, necesito hacer una **radiografía inicial** del archivo crudo (`bank-additional.csv`). Trabajar a ciegas es el mayor error en Data Science. Este paso me permitirá detectar problemas de calidad de datos, formatos incorrectos y valores nulos para trazar un plan de limpieza estructurado.

## 📂 Origen de los Datos y Metodología
* **Fuente:** Dataset en crudo almacenado localmente en `../data/bank-additional.csv`.
* **Herramienta Custom:** Para mantener este notebook limpio y profesional, he empaquetado todo el código repetitivo de exploración en mi propia librería `src.sp_eda`. 

**Pasos de esta auditoría automática:**
1. Muestra aleatoria (para entender la semántica de los datos).
2. Dimensiones y tipos de datos (auditoría de memoria y formato).
3. Conteo exacto de valores nulos y duplicados.
4. Resumen estadístico y análisis de frecuencias para detectar anomalías (ej. errores tipográficos o de codificación).


In [1]:
# -------------------------------------------------------------------------
# PREPARACIÓN DEL ENTORNO Y LIBRERÍAS
# -------------------------------------------------------------------------
import pandas as pd
import sys

# Configuración de Pandas para visualización completa
pd.set_option('display.max_columns', None) 

# Añado la raíz del proyecto al PATH para poder importar mis propios módulos.
# Esto es una buena práctica para estructurar proyectos complejos.
sys.path.append('../')

# Importo mi módulo personalizado de Análisis Exploratorio
from src import sp_eda as se

print("✅ Entorno configurado y módulo 'sp_eda' cargado correctamente.")


✅ Entorno configurado y módulo 'sp_eda' cargado correctamente.


In [2]:
# -------------------------------------------------------------------------
# CARGA DE DATOS Y EJECUCIÓN DEL EDA
# -------------------------------------------------------------------------
ruta_crudos = "../data/bank-additional.csv"

try:
    df_crudo = pd.read_csv(ruta_crudos)
    print("✅ Datos cargados con éxito. Iniciando auditoría...\n" + "="*50)
    
    # Ejecuto la función principal de mi módulo
    se.eda_preliminar(df_crudo)
    
except FileNotFoundError:
    print(f"❌ Error: No se ha encontrado el archivo en la ruta {ruta_crudos}")


✅ Datos cargados con éxito. Iniciando auditoría...
--- 1. MUESTRA ALEATORIA ---


,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
3306,3403,35.0,admin.,MARRIED,university.degree,0.0,0.0,0.0,telephone,420,4,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,86",5191,no,29-octubre-2015,40.592,-68.430,63f971fc-819c-48de-b5a9-2e619d167df4
37427,37594,62.0,management,DIVORCED,high.school,0.0,0.0,0.0,cellular,405,1,999,0,NONEXISTENT,-2.9,"92,201","-31,4","0,861","5076,2",yes,29-septiembre-2015,26.834,-123.387,afa34393-cdf2-43a3-9fbd-a461259bfb12
22325,22422,53.0,technician,MARRIED,professional.course,0.0,0.0,0.0,cellular,136,1,999,0,NONEXISTENT,1.4,"93,444","-36,1","4,964","5228,1",no,23-enero-2019,35.790,-119.107,609d3f4b-2ea0-43da-beb3-58dc76e9de9c
22917,23014,55.0,services,MARRIED,high.school,NaN,1.0,0.0,cellular,209,3,999,0,NONEXISTENT,1.4,"93,444","-36,1","4,965","5228,1",no,4-diciembre-2018,39.284,-84.089,6295d85d-df03-46e8-a86f-775053eb9c02
5142,5239,33.0,blue-collar,MARRIED,basic.9y,0.0,1.0,0.0,telephone,50,3,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,14-abril-2017,30.682,-80.213,304a068f-877f-4ef2-b979-f4f94f0cb23f



--- 2. DIMENSIONES ---
Filas: 43000 | Columnas: 24

--- 3. INFO GENERAL ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  object 
 16 

,% Nulos
euribor3m,21.525581
default,20.886047
age,11.906977
education,4.202326
loan,2.386047
housing,2.386047
cons.price.idx,1.095349
job,0.802326
date,0.576744
marital,0.197674



--- 5. DUPLICADOS ---
✅ No hay filas duplicadas.

--- 6. ESTADÍSTICAS NUMÉRICAS ---


,count,mean,std,min,25%,50%,75%,max
Unnamed: 0,43000.0,20651.099047,11868.770913,0.000,10381.75000,20642.5000,30930.25000,41187.000
age,37880.0,39.977112,10.437957,17.000,32.00000,38.0000,47.00000,98.000
default,34019.0,0.000088,0.009390,0.000,0.00000,0.0000,0.00000,1.000
housing,41974.0,0.535998,0.498708,0.000,0.00000,1.0000,1.00000,1.000
loan,41974.0,0.155620,0.362499,0.000,0.00000,0.0000,0.00000,1.000
duration,43000.0,257.739279,258.666033,0.000,102.00000,179.0000,319.00000,4918.000
campaign,43000.0,2.567233,2.772294,1.000,1.00000,2.0000,3.00000,56.000
pdays,43000.0,962.330953,187.260394,0.000,999.00000,999.0000,999.00000,999.000
previous,43000.0,0.174023,0.497366,0.000,0.00000,0.0000,0.00000,7.000
emp.var.rate,43000.0,0.077128,1.573898,-3.400,-1.80000,1.1000,1.40000,1.400



--- 7. COLUMNAS DE TEXTO (Top 10 valores) ---

-> Columna: JOB (Total únicos: 11)


,Valor,Frecuencia
0,admin.,10873
1,blue-collar,9654
2,technician,7026
3,services,4162
4,management,3050
5,retired,1790
6,entrepreneur,1522
7,self-employed,1489
8,housemaid,1123
9,unemployed,1063


   (Se muestran solo los 10 primeros de 11)

-> Columna: MARITAL (Total únicos: 3)


,Valor,Frecuencia
0,MARRIED,25999
1,SINGLE,12105
2,DIVORCED,4811



-> Columna: EDUCATION (Total únicos: 7)


,Valor,Frecuencia
0,university.degree,12722
1,high.school,9925
2,basic.9y,6309
3,professional.course,5477
4,basic.4y,4356
5,basic.6y,2386
6,illiterate,18



-> Columna: CONTACT (Total únicos: 2)


,Valor,Frecuencia
0,cellular,27396
1,telephone,15604



-> Columna: POUTCOME (Total únicos: 3)


,Valor,Frecuencia
0,NONEXISTENT,37103
1,FAILURE,4461
2,SUCCESS,1436



-> Columna: CONS.PRICE.IDX (Total únicos: 26)


,Valor,Frecuencia
0,"93,994",7938
1,"93,918",6937
2,"92,893",5985
3,"93,444",5349
4,"94,465",4522
5,"93,2",3731
6,"93,075",2552
7,"92,201",803
8,"92,963",742
9,"92,431",471


   (Se muestran solo los 10 primeros de 26)

-> Columna: CONS.CONF.IDX (Total únicos: 26)


,Valor,Frecuencia
0,"-36,4",8020
1,"-42,7",7004
2,"-46,2",6057
3,"-36,1",5408
4,"-41,8",4568
5,-42,3782
6,"-47,1",2581
7,"-31,4",815
8,"-40,8",748
9,"-26,9",477


   (Se muestran solo los 10 primeros de 26)

-> Columna: EURIBOR3M (Total únicos: 309)


,Valor,Frecuencia
0,"4,857",2287
1,"4,962",2124
2,"4,963",2019
3,"4,961",1594
4,"4,964",964
5,"4,856",961
6,"1,405",954
7,"4,864",871
8,"4,965",866
9,"4,96",836


   (Se muestran solo los 10 primeros de 309)

-> Columna: NR.EMPLOYED (Total únicos: 11)


,Valor,Frecuencia
0,"5228,1",16980
1,"5099,1",8937
2,5191,8020
3,"5195,8",3830
4,"5076,2",1748
5,"5017,5",1133
6,"4991,6",812
7,"5008,7",683
8,"4963,6",665
9,"5023,5",181


   (Se muestran solo los 10 primeros de 11)

-> Columna: Y (Total únicos: 2)


,Valor,Frecuencia
0,no,38156
1,yes,4844



-> Columna: DATE (Total únicos: 1825)


,Valor,Frecuencia
0,28-febrero-2018,95
1,28-febrero-2016,94
2,28-febrero-2015,92
3,28-febrero-2017,89
4,28-febrero-2019,85
5,30-abril-2018,58
6,30-septiembre-2017,56
7,30-noviembre-2017,55
8,30-junio-2015,55
9,30-abril-2015,50


   (Se muestran solo los 10 primeros de 1825)

-> Columna: ID_ (Total únicos: 43000)


,Valor,Frecuencia
0,993bbbd6-4dbc-4a40-a408-f91f8462bee6,1
1,089b39d8-e4d0-461b-87d4-814d71e0e079,1
2,e9d37224-cb6f-4942-98d7-46672963d097,1
3,3f9f49b5-e410-4948-bf6e-f9244f04918b,1
4,9991fafb-4447-451a-8be2-b0df6098d13e,1
5,eca60b76-70b6-4077-80ba-bc52e8ebb0eb,1
6,d63ede72-0b6d-45b1-8872-385ac6897f65,1
7,5e3483e5-236d-437d-8351-541f9d09b9dd,1
8,9ebb7576-92de-4949-b2e3-e56890740112,1
9,85e7d06a-50c4-4b82-bcb7-45525c7f7f6a,1


   (Se muestran solo los 10 primeros de 43000)


## 💡 Diagnóstico y Plan de Acción (Insights del EDA)

Tras analizar el reporte generado por mi módulo, he detectado varios problemas críticos de calidad de datos que deberé solucionar en la fase de ETL (Siguiente Notebook). Esta es mi hoja de ruta:

### 🚨 1. Problemas de Formato (Tipos de datos incorrectos)
* **Números guardados como texto (Object):** Columnas como `cons.price.idx`, `cons.conf.idx`, `euribor3m` y `nr.employed` deberían ser numéricas (float). Se han guardado como texto porque **usan una coma decimal (`,`) en lugar de un punto (`.`)**. Tendré que reemplazar esto y convertirlas.
* **Fechas como texto:** La columna `date` contiene texto mezclando números y meses en español (ej. *18-diciembre-2015*). Necesita ser parseada a formato `datetime`.

### 🚨 2. Inconsistencias en Texto (Mayúsculas y minúsculas)
* Al observar las frecuencias, veo que `marital` está todo en MAYÚSCULAS ('MARRIED'), mientras que `job` está en minúsculas ('admin.'). 
* Hay caracteres extraños y puntuación en variables categóricas (ej. `admin.`, `blue-collar`). 
* **Plan:** Estandarizaré todo el dataset a minúsculas y reemplazaré guiones y puntos por barras bajas (`_`) para tener consistencia.

### 🚨 3. Gestión de Nulos
Tenemos una cantidad significativa de valores perdidos:
* `euribor3m` (~21.5%) y `default` (~20.8%) tienen un gran volumen de nulos.
* `age` (~11.9%) también falta bastante.
* **Plan:** En el próximo notebook evaluaré si imputar estos valores (con la mediana o moda) o si crear una categoría 'desconocido'.

### 🚨 4. Estandarización de Columnas
* La columna `Unnamed: 0` parece ser un índice residual del CSV, la eliminaré.
* Los nombres de las columnas usan puntos (`emp.var.rate`). Aplicaré una función de mi módulo de limpieza para pasarlos a formato `snake_case` (`emp_var_rate`).

**Próximo Paso 👉:** Pasar al **Notebook 02** de Limpieza de Datos, donde aplicaré mi módulo `src.sp_lim.py` para solucionar todos estos hallazgos estructurales.
